node/lib/util/rm.js

```diff
@@ -0,0 +1,482 @@
+ * @param this.removedSubmodules (quasi-out parameter)
+ * @param this.removedFiles (quasi-out parameter)
+ * @return 1 always
+ */
+const prepRootLevelChange = function(path) {
+    const indexEntry = this.index.getByPath(path);
+    if (FILEMODE.COMMIT === indexEntry.mode) {
+        if (this.cached && path in this.metaRepoStatus.submodules) {
+            this.badFiles.push(path);
+        }
+        else {
+            this.removedSubmodules.push(path);
+        }
+    }
+    else {
+        if (this.cached && this.staged.has(path) ||
+            this.workdir.has(path)) {
+            this.badFiles.push(path);
+        }
+        else {
+            this.removedFiles.push(path);
+        }
+    }
+    return 1;
+};
+
+const checkFileForLocalChanges = function(pathToRemove) {
+    let bad;
+    if (this.cached) {
+        if (!this.staged.has(pathToRemove)) {
+            bad = false;
+        }
+        else {
+            this.indexFilesToCheck.push(pathToRemove);
+        }
+    }
+    else {
+        bad = this.staged.has(pathToRemove) ||
+            this.workdir.has(pathToRemove);
+    }
+
+    const metaPath = path.join(this.pathFromRoot, pathToRemove);
+    if (bad) {
+        this.badFiles.push(metaPath);
+    }
+    else {
+        this.removedFiles.push(metaPath);
+    }
+
+    return 1;
+};
+
+/**
+ * Remove the specified `paths` in the specified `repo`.  If a path in
+ * `paths` refers to a file, remove it.  If it refers to a directory,
+ * and recursive is true, remove it recursively.  If it's false, throw
+ * a UserError.  If a path to be removed does not exist in the index,
+ * throw a UserError.  If a path to be removed is not clean, and force
+ * is false, throw a UserError.
+ *
+ * If --cached is supplied, the paths will be removed from the index
+ * but not from disk.
+ *
+ * If there are any dirty paths, the meta or submodule indexes may be
+ * dirty upon return from this function.  Unfortunately due to a
+ * nodegit bug, we can't reload the index:
+ * https://github.com/nodegit/nodegit/issues/1478
+ *
+ * @async
+ * @param {NodeGit.Repository} repo
+ * @param {String []}          paths
+ * @param {Boolean}            recursive
+ * @param {Boolean}            cached (remove from index not disk)
+ * @param {Boolean}            force
+ */
+exports.rmPaths = co.wrap(function *(repo, paths, recursive, cached, force) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.isArray(paths);
+    assert.isBoolean(recursive);
+    assert.isBoolean(cached);
+    assert.isBoolean(force);
+
+    //note that "submodule" here is something of a misnomer in case 4
+    const bySubmodule = new Map();
+    const index = yield repo.index();
+    for (const path of paths) {
+        const spec = yield exports.getRemovalSpec(repo, index, path,
+                                                  recursive);
+        const submodule = spec.pathFromRoot;
+        if (submodule in bySubmodule) {
```

Comment:
I'm a little surprised that `in` works; I think you'd want to use the `has` method with `Map`.